### PR TITLE
Organize utils.

### PR DESCRIPTION
**Pull request by knighton**
Break up util.py: `importing` `merging` `retrying` `shared` `shorthand`.

### CODE REVIEW

File: streaming/storage/download.py [hunk 1]
```diff
@@ -474,18 +475,25 @@ def download_file(remote: Optional[str], local: str, timeout: float):
...
-        elapsed = time() - t0
-        if timeout < elapsed:
-            raise TimeoutError(
-                f'Waited longer than {timeout}s for other worker to download {local}.')
-        sleep(0.25)
+    start_time = time()
+    while True:
+        sleep(poll_interval)
+        if os.path.exists(filename):
+            sleep(poll_interval)
```
Comment (karan6181) [inline streaming/storage/download.py:495]: Do we need to sleep here since the file already exists now?

File: streaming/storage/download.py [hunk 2]
```diff
@@ -474,18 +475,25 @@ def download_file(remote: Optional[str], local: str, timeout: float):
         download_from_local(remote, local)
 
 
-def wait_for_download(local: str, timeout: float = 60) -> None:
-    """Wait for a download by another thread/process to complete.
+def wait_for_file_to_exist(filename: str, poll_interval: float, timeout: float,
```
Comment (karan6181) [inline streaming/storage/download.py:478]: Can we keep this method in util directory? Looks like this method does not fit well here with other cloud provider download methods.

File: streaming/util/importing.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    """Get import exception message.
+
+    Args:
+        package_name (str): Package name.
+
+    Returns:
+        str: Exception message.
+    """
+    return f'Streaming was installed without {package_name} support. ' + \
+            f'To use {package_name} related packages with Streaming, run ' + \
```
Comment (karan6181) [inline streaming/util/importing.py:24]:
```suggestion
           f'To use {package_name} related packages with Streaming, run ' + \
```

File: streaming/util/importing.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+            f'`pip install \'mosaicml-streaming[{package_name}]\'`.'
```
Comment (karan6181) [inline streaming/util/importing.py:25]:
```suggestion
           f'`pip install \'mosaicml-streaming[{package_name}]\'`.'
```

File: streaming/storage/download.py [hunk 5]
```diff
@@ -474,18 +475,25 @@ def download_file(remote: Optional[str], local: str, timeout: float):
...
-        sleep(0.25)
+    start_time = time()
+    while True:
+        sleep(poll_interval)
+        if os.path.exists(filename):
+            sleep(poll_interval)
+            break
+        dt = time() - start_time
+        if dt > timeout:
+            raise RuntimeError(f'{err_msg}' + f'{timeout:.3f} < {dt:.3f} secs.')
```
[the remark this answers is not in the document]
Comment (snarayan21) [reply streaming/storage/download.py:499]: +1

File: streaming/util/importing.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+
+def redirect_imports(new_fqdn: str) -> None:
```
Comment (snarayan21) [inline streaming/util/importing.py:28]: has this method been tested? just want to make sure that the indexing into `inspect.stack()` and so on to get the current module is working as intended

File: streaming/util/importing.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    Args:
+        new_fqdn (str): Fully-qualified dot-separated target module path.
+    """
+    frame = inspect.stack()[1]
+    module = inspect.getmodule(frame[0])
+    if module is None:
+        raise RuntimeError('Module was None.')
+    old_fqdn = module.__name__
+
+    # old = sys.modules[old_fqdn]
```
Comment (snarayan21) [inline streaming/util/importing.py:40]: nit: delete comment or make this more descriptive?

File: streaming/storage/download.py (hunk 2 above)
Comment (snarayan21) [reply streaming/storage/download.py:478]: should this maybe go in `retrying.py`?

**Review by snarayan21 — COMMENTED** [review]
minor comments!

File: streaming/storage/download.py (hunk 1 above)
Comment (knighton) [reply streaming/storage/download.py:495]: Yeah that is mysterious... from streaming experience, I wonder if this line is for a race condition where a small file is getting written by someone else. Would prefer to be conservative for the time being...

File: streaming/storage/download.py (hunk 2 above)
Comment (knighton) [reply streaming/storage/download.py:478]: This move portends future PRs that encapsulate the complexity of downloading with size check/hash validation and other "smart" functionality within the storage APIs, simplifying StreamingDataset proper -- wdyt in this context?

File: streaming/storage/download.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (knighton) [reply streaming/storage/download.py:499]: done

File: streaming/util/importing.py (hunk 6 above)
Comment (knighton) [reply streaming/util/importing.py:28]: added test

File: streaming/storage/download.py (hunk 2 above)
Comment (knighton) [reply streaming/storage/download.py:478]: Would prefer to punt on messing with this to after future PRs are in and have fleshed things out?

**Pull request merged**